### DISCUSSION TITLE
UID Assignment?

### DISCUSSION THREAD

**Discussion by bitsybabe** [orchetect/swift-midi#185] [Q&A]
How are MIDI UIDs assigned in your kit? If I connect device A with In and Out ports via bluetooth MIDI and store its UID in my app, and then connect the same device through a Scarlett 2i4 using 5-pin wires, will it get a different UID?  I'm assuming yes.

If I then use the bluetooth jack again, will it reacquire its original UID? Or is that going to be regenerated and come up with something different?

And - is there any chance that if device B is connected by moving the bluetooth jack from device A, it'll get device A's UID?

I'm writing an app that helps users manage their collection of synthesizers, within which I am assuming (like myself) they'll have more than 1 of a physical device (e.g., one for gigging and one for the studio).  So, I'm allotting space in the data model to store as many MIDI UIDs as I need to, associated with a named physical device. I just need to understand if they're worth saving or I need to create UI for the user to match them up.

**Comment by orchetect** [comment]
> If I connect device A with In and Out ports via bluetooth MIDI and store its UID in my app, and then connect the same device through a Scarlett 2i4 using 5-pin wires, will it get a different UID? I'm assuming yes.

More than likely yes, just considering the change of medium. You can test this of course. Pop open the MIDISystemInfo demo app and see what its unique ID is each time the device is attached/connected, for each connection method.

> If I then use the bluetooth jack again, will it reacquire its original UID?

It should, but again, you can check this for yourself to be sure. This would be handled by the system / Core MIDI Bluetooth driver.

> a named physical device

Since you're seemingly concerned with devices, you could retain the device's unique ID, and re-scan its endpoints to see its in/out configuration and refer to its inputs or outputs as offsets/indexes (vis-a-vis an inputs array, and outputs array). It depends on what functionality you're after. Otherwise you could retain individual endpoint unique IDs instead, if they need to be referenced specifically in your model.

Typically If you're creating connections with [`addInputConnection()`](https://orchetect.github.io/MIDIKit/documentation/midikit/midimanager/addinputconnection(tooutputs:tag:mode:filter:receiver:)-5xxyz)/[`addOutputConnection()`](https://orchetect.github.io/MIDIKit/documentation/midikit/midimanager/addoutputconnection(toinputs:tag:mode:filter:)-3a56s), using [`.uniqueID()`](https://orchetect.github.io/MIDIKit/documentation/midikit/midiendpointidentity/uniqueid(_:)) to identify each endpoint to connect to would (and should) be sufficient. But if you find that IDs change for the same device between Bluetooth and MIDI-DIN/USB connection, there is also [`.uniqueIDWithFallback()`](https://orchetect.github.io/MIDIKit/documentation/midikit/midiendpointidentity/uniqueidwithfallback(id:fallbackdisplayname:)) which can fall back to identifying an endpoint by its Display Name property (not Name property) in the event the unique ID is not found in the system.

**Reply by orchetect** [reply]
If there's any room for clarification of this in the library or documentation let me know.

One of the to-do items is to add an in-depth guide / FAQ on how MIDI works and nuances of how different platforms and connections interact.
